enaml/core/parser/base_parser.py

```diff
@@ -779,13 +779,25 @@ def p_operator_expr2(self, p):
     def p_operator_expr3(self, p):
         ''' operator_expr : DOUBLECOLON suite '''
         lineno = p.lineno(1)
-        mod = ast.Module()
-        mod.body = p[2]
-        for item in ast.walk(mod):
+
+        for item in ast.walk(ast.Module(body=p[2])):
             if type(item) in self._NOTIFICATION_DISALLOWED:
                 msg = '%s not allowed in a notification block'
                 msg = msg % self._NOTIFICATION_DISALLOWED[type(item)]
                 syntax_error(msg, FakeToken(p.lexer.lexer, item.lineno))
+
+        func_node = ast.FunctionDef()
+        func_node.name = 'f'
+        func_node.args = self._make_args([])
+        func_node.decorator_list = []
+        if IS_PY3:
+            func_node.returns = None
+        func_node.body = p[2]
+        func_node.lineno = lineno
+
+        mod = ast.Module(body=[func_node])
+        ast.fix_missing_locations(mod)
+
```

Comment:
I don't think this is correct. Can you give an example of the closure variables you're hoping to pick up with this change?

Also, I don't think this is going to work as you expect. When the `::` operator is run to generate the expression handler, it wraps the generated bytecode in it's own function, which is then invoked by the handler: https://github.com/nucleic/enaml/blob/master/enaml/core/operators.py#L70  So each time the notification is triggered, this new internal function is going to be defined. I'm not entirely sure how this is passing tests.

Reply:
But your example already works without this change, because all `LOAD_GLOBALS` are re-written to `LOAD_NAME`.

Reply:
Can you give me an example of something that doesn't work currently, that this change fixes? Also, can you link me to where you "extract the body of the function and pass it forward"?

Reply:
Ah, okay, you mean that doesn't work on Python 3. Sorry.

We just need to recursively rewrite LOAD_GLOBAL to LOAD_NAME for nested code objects.

Reply:
It is not so simple because LOAD_NAME fails to access fast locals of the outer scope. I will answer in more details later. I am attending a conference this week.

Reply:
I don't think there's anything in fast locals that the inner code will need to access.

Reply:
Still looking at this kind of comprehension:
```python
a = 1
[a for i in range (2)]
```

Digging deeper I found that:
- when Python compile the code as if it was a top-level module (which it does without this fix), it uses STORE_GLOBAL and LOAD_GLOBAL to handle `a`
- to unify the handling of comprehensions between operators and declarative functions I refactored the handling of global variables such that as LOAD_GLOBAL and STORE_GLOBAL are replaced by LOAD_NAME/STORE_NAME save if the variable is marked using the `global` keyword, both in declarative functions and operators. Currently this is not done if there is no comprehension in an operator which is inconsistent. cf https://github.com/nucleic/enaml/blob/master/enaml/core/compiler_common.py#L685
- finally when creating the operator (https://github.com/nucleic/enaml/blob/master/enaml/core/operators.py#L156), we optimize the code for locals (https://github.com/nucleic/enaml/blob/master/enaml/core/operators.py#L49) which replaces LOAD_NAME/STORE_NAME for local variables by LOAD_FAST/STORE_FAST.

As a consequence a is not stored in the namespace, which explains the failure.

From my point of view the proposed fixed is the more correct since the body of a notification operator is truly a function body (even without return and yield). Alternatively we could suppress the global variable analysis in operators. 

I will update the code to always analyse global variables in operators for the time being, so that the current behavior is consistent.